doc/source/development/rfc/rfc104_gdal_cli.rst

```diff
@@ -0,0 +1,1175 @@
+
+- a pointer to a variable of the type consistent with its type (``bool*``,
+  ``std::string*``, ``int*``, ``double*``, ``GDALArgDatasetValue``,
+  ``std::vector<std::string>*``, ``std::vector<int>*``, ``std::vector<double>*``,
+  ``std::vector<GDALArgDatasetValue>*``). The value pointed by this pointer is
+  modified by the ``Set()`` methods.
+
+- an optional declared default value (if no declared default value, nor explicitly set value, the initial state of the variable pointed by the above mentioned pointer will be the value, that is a kind of implicit default value)
+
+- whether it has been explicitly set.
+
+- ``key: list of string`` free metadata. For example the ``format`` argument
+  uses the ``required_capabilities`` key and ``DCAP_RASTER``, ``DCAP_VECTOR``,
+  ``DCAP_CREATE``, etc. as values to declare the expected capability of the
+  specified input/output format.
+
+- a list of action callbacks that are triggered by the ``Set`` method.
+
+- a list of validation callbacks that are triggered by the ``Set`` method.
+  Each callback may return false to mean that the new value is invalid.
+
+GDALArgDatasetValue
+*******************
+
+This class models an argument that holds a GDALDataset. It stores a dataset
+name or a ``GDALDataset*`` pointer itself (with information if its ownership
+is transferred to the GDALArgDatasetValue instance).
+This is done this way to be compatible both of CLI usage where only dataset names
+are specified, or programmatic usages where passing dataset names or passing/getting
+``GDALDataset*`` pointers might be preferred. That later mode is also used by
+``gdal vector pipeline`` to bind together the output of a step to the input of
+the following step.
+
+``GDALArgDatasetValue`` has also ``inputFlags`` and ``outputFlags`` properties
+to indicate if it supports specifying only the dataset name, only the dataset object,
+or both when it is used as an input, or if it generates the dataset name,
+the dataset object or both when it is used as an output.
+
+The GDALAlgorithm class itself has logic, triggered during the validation phase,
+to open the ``GDALDataset*`` from its name for input arguments, taking into
+account potential open options and allowed input formats.
+It has also very specific logic to realize that if both the ``input`` and
+``output`` arguments point to the same dataset name, a single ``GDALDataset*``
+instance must be set onto both for drivers that require it (typically SQLite
+based ones).
+
+GDALAlgorithmRegistry
+*********************
+
+Instances of this class store a list of C++ types implementing GDALAlgorithm.
+
+It is used by GDALAlgorithm itself for nodes to reference their potential children.
+
+It is extended by a GDALGlobalAlgorithmRegistry to offer a singleton that lists all
+top-level nodes (``raster``, ``vector``, etc.). Potentially code external to GDAL
+could register a new command available for use by :program:`gdal` in a GDAL plugin.
```

Comment:
"Register new commands by a plugin". Isn't it too flexible? It may open security problems.

Reply:
not more than having GDAL driver plugins. There isn't anything special done in the code to allow registration of a new command by a plugin. And I don't see how that could be prevented (apart from using the existing mechanisms to disallow driver loading at build time or runtime)